src/createAsyncStoragePersister/asyncThrottle.ts

```diff
@@ -0,0 +1,56 @@
+export interface AsyncThrottleOptions {
+  interval?: number
+  onError?: (error: unknown) => void
+}
+
+const noop = () => {}
+
+export function asyncThrottle<Args extends readonly unknown[]>(
+  func: (...args: Args) => Promise<void>,
+  { interval = 1000, onError = noop }: AsyncThrottleOptions = {}
+) {
+  if (typeof func !== 'function') throw new Error('argument is not function.')
+
+  let running = false
+  let lastTime = 0
+  let timeout: ReturnType<typeof setTimeout>
+  let currentArgs: Args | null = null
+
+  const execFunc = async () => {
+    if (currentArgs) {
+      const args = currentArgs
+      currentArgs = null
+      try {
+        running = true
+        await func(...args)
+      } catch (error) {
+        onError(error)
+      } finally {
```

Comment:
hm, I don't fully understand why we would need the `onError` callback. If remove the `catch` statement here and just let the error bubble up to the call side, what would happen?

Reply:
Remove `catch` directly, keep `try...finally` is fine for `asyncThrottle`. But then error handling must be handled internally by the `func` itself. `asyncThrottle` returns a non-waitable function.

If we decide to handle errors inside the `func`, I'll change it right away.

Reply:
i think it's expected that if `asyncThrottle` returns a `Promise` that it can fail

Reply:
Sorry, I wasn't clear.

If the `onError` callback is retained, the error handling will look like this：
```typescript
return {
    persistClient: asyncThrottle(
      (persistedClient: PersistedClient) => storage.setItem(key, serialize(persistedClient)),
      { 
        interval: throttleTime,
        onError: (err: unknown) => {
          // handing error here
        } 
      }
    ),
// ....
```

If the `onError` callback is removed directly：
```typescript
return {
    persistClient: asyncThrottle(
      async (persistedClient: PersistedClient) => {
        try {
          storage.setItem(key, serialize(persistedClient))
        } catch (error) {
          // handing error here
        }
      }, 
      throttleTime
    ),
// ....
```



> i think it's expected that if `asyncThrottle` returns a `Promise` that it can fail

`asyncThrottle` returns a non-waitable function. Do you mean to return a function that returns a `promise`?
If so, this is strange. `asyncThrottle` returns a function that may be called 10 times， but the 'func' may be executed only twice. `promise` either resolves or rejects, and cannot be left dangling.
And the code will be more complex than now, I still think the previous two better a little , do you think?
☺

Reply:
I see, my mistake. let's keep the callback then 👍